### ISSUE TITLE
Nullness: upper bounds for j.u.Optional and j.l.Class

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1098]
Take this simple example:

```
import java.util.Optional;

class OptionalDemo {
  <T> void opt(Optional<T> p1, T p2) {}
  <T> void cls(Class<T> p1, T p2) {}

  void use() {
    opt(Optional.empty(), null);
    cls(this.getClass(), null);
  }
}
```

The declaration of method `opt` doesn't pass the Nullness Checker, because the upper bound of T is incompatible with the non-null bound of Optional.
For j.l.Class we used nullable as upper bound, conveniently allowing the `cls` declaration.
The invocation of method `cls` is however forbidden, which doesn't seem desirable.

For both Optional and Class one can make the argument that the nullity of the type argument doesn't really matter and that any instantiation is fine.

Attached a possible change to the signature of Optional that works conveniently in more code.

[Optional-Nullable.zip](https://github.com/typetools/checker-framework/files/815407/Optional-Nullable.zip)

**Comment by mernst** [comment]
I agree that the inconsistency between `Optional` and `Class` is undesirable.

I would be OK with permitting unbounded type variables as in `Optional<T>`.  However, I would prefer that programmers do not write `Optional<@Nullable String>` which is misleading to readers.  Is there any way to achieve that?

We'll eventually need more annotations than those in your diff.
Consider the `Optional.ifPresent` routine:
```java
  public void ifPresent(Consumer<? super T> consumer) {
```

It's currently commented out because the `Consumer` class is defined in JDK 8.
However, we would want to change `T` to `@NonNull T` so that the signatures of these routines are the same:
```java
Optional<@NonNull String>.ifPresent()
Optional<@Nullable String>.ifPresent()
```

**Comment by mernst** [comment]
Making the change permits the examples in this issue to type-check, but it causes other previously-passing code to fail to type-check under the nullness type system.  Here is an example:

```
    void m(Date d) {
        Class<? extends Date> c = d.getClass();
    }
```

```
error: [assignment.type.incompatible] incompatible types in assignment.
        Class<? extends Date> c = d.getClass();
                                            ^
  found   : @Initialized @NonNull Class<? extends @Initialized @Nullable Date>
  required: @UnknownInitialization @Nullable Class<? extends @Initialized @NonNull Date>
```

Do we have evidence about which type of code is more common, and therefore more desirable to support while forcing user annotations for the other?

**Comment by wmdietlGC** [comment]
I think the Java compiler handles `Object.getClass` specially.
Our signature at https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/lang/Object.java#L19 says that the return type has a non-null upper bound.
Note how the upper bound depends on the receiver type, which is not clear from the signature.

I had a quick look, but didn't see special handling for `Object.getClass` in the Nullness Checker or anywhere else. Maybe we need to add such logic to ensure that the wildcard returned from javac is `@NonNull`.

**Comment by smillst** [comment]
The Checker Framework special cases `getClass` in [AnnotatedTypeFactory#adaptGetClassReturnTypeToReceiver](https://checkerframework.org/api/org/checkerframework/framework/type/AnnotatedTypeFactory.html#adaptGetClassReturnTypeToReceiver-org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType-org.checkerframework.framework.type.AnnotatedTypeMirror-).  So we could change that code such that the wildcard is `@NonNull`.

**Comment by wmdietlGC** [comment]
Thanks for tracking down this method @smillst ! I had thought that we did this somewhere but didn't find it...
Shouldn't this logic already pick up the `@NonNull` from https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/lang/Object.java#L19 ?

**Comment by mernst** [comment]
I tried using adaptGetClassReturnTypeToReceiver to make the return value of getClass always be an upper-bounded wildcard, per @smillst's suggestion.  Here is the code:
https://github.com/typetools/checker-framework/pull/1374/commits/f9658b7a4bf31ba050256ea60bda8add4c8c86e2

That made the

```
    void m(Date d) {
        Class<? extends Date> c = d.getClass();
    }
```

example typecheck, but the

```    cls(this.getClass(), null);```

example stopped type-checking.  Is there a way to make both use cases typecheck?  Or does my code need to examine the context?

**Comment by smillst** [comment]
What about marking the type parameter with [@Covariant](https://checkerframework.org/manual/#covariant-type-parameters)?

**Comment by smillst** [comment]
Fixed via https://github.com/typetools/checker-framework/commit/a9c2179ec2ab3918ac6c6f6e66d2f10ccaf35978.
